### ISSUE TITLE
[Feature] Service Discovery using Consul

### ISSUE THREAD

**Issue by 0xbaadf00d** [mholt/caddy#1630]
Hello,

It could be very nice if Caddy can be dynamically reconfigured by Consul ([consul.io](https://www.consul.io/)). 

In a micro-services environment (or not) it allow caddy to get list of available proxy address and path without manual reload.

Nginx have this kind of feature, but only on the "Plus" version.  I think it was a killer feature for Caddy.

https://www.nginx.com/blog/service-discovery-with-nginx-plus-and-consul/
https://www.nginx.com/blog/service-discovery-nginx-plus-srv-records-consul-dns/

Thanks

**Comment by francislavoie** [comment]
In the short-term, check out Traefik. That's what it specializes in. Maybe a plugin would be more appropriate for this feature?

**Comment by 0xbaadf00d** [comment]
Yes i think a plugin could  be a good solution. But is plugin can create in-memory caddyfile configuration ? Dont Know if a hook Exist...

**Comment by mholt** [comment]
I think the rewrite of the proxy will make this possible. (See linked issue.)

**Comment by josegonzalez** [comment]
Semi-related: https://github.com/brimstone/caddy-consul

**Comment by mholt** [comment]
@josegonzalez That's cool, what does it do exactly? (ELI5-and-have-never-used-consul)

**Comment by josegonzalez** [comment]
Seems like a caddy plugin (not that I've used any) that will:

- watch for changes in the consul catalog of services
- build a caddyfile config
- reload caddy

**Comment by krishamoud** [comment]
Hi everyone,  I forked https://github.com/brimstone/caddy-consul for my own use.

https://github.com/krishamoud/caddy-consul is available for anyone who wants to give it a shot.  I tried to make it as generic as possible so you can add your own template and the caddyfile will look how you want it to.

**Comment by mholt** [comment]
Cool. I think this is now possible by reverse proxying to `srv://` endpoints. https://github.com/mholt/caddy/pull/1915

**Comment by josegonzalez** [comment]
Note: For those more familiar with consul, a slightly faster way would be to watch for consul key changes on the local client via stale reads (non-stale reads will go direct to the consul cluster leader). Consul can then invalidate records for services you are watching for _before_ you actually invalidate DNS.

**Comment by mholt** [comment]
@josegonzalez How trivial is that change to implement, do you think? Is it something that could be a good PR?

**Comment by tcurdt** [comment]
I am a little confused on the status of this. The issues is closed but I cannot seem to find any docs on a consul integration. Is the following traefik config possible with caddy or not?

```
        image = "traefik:v2.4"
        args = [
          # consul
          "--providers.consulCatalog=true",
          "--providers.consulCatalog.prefix=traefik",
          "--providers.consulCatalog.exposedByDefault=false",
          # ports
          "--entrypoints.web.address=:80",
          "--entrypoints.websecure.address=:443",
          "--entrypoints.websecure.http.tls=true",
          # redirects
          "--entrypoints.web.http.redirections.entrypoint.to=websecure",
          "--entrypoints.web.http.redirections.entrypoint.scheme=https",
          # certificats
          "--entrypoints.websecure.http.tls.certResolver=le",
          "--certificatesresolvers.le.acme.email=ops@foo.com",
          "--certificatesresolvers.le.acme.storage=/acme.json",
          "--certificatesresolvers.le.acme.tlschallenge=true",
```
```
        "traefik.enable=true",
        "traefik.http.routers.backend.tls=true",
        "traefik.http.routers.backend.entrypoints=websecure",
        "traefik.http.services.backend.loadbalancer.server.port=${NOMAD_PORT_nodejs}",
        "traefik.http.routers.backend.rule=Host(`${var.domain}`) && PathPrefix(`/api`)",
```

**Comment by josegonzalez** [comment]
@mholt forgot to respond but the big thing with my method is you would need a method to store global state (so that you can store the lookup of services on your end). Not sure how I'd go about that with caddy, but definitely could be a good optimization (maybe a scheme of `consul://`?).

**Comment by francislavoie** [comment]
@josegonzalez we're working on a rewrite of upstream management which will bring major improvements to SRV upstreams. I think this will help significantly with Consul users as well: https://github.com/caddyserver/caddy/pull/4470
